aten/src/ATen/CPUGenerator.cpp

```diff
@@ -1,49 +1,195 @@
 #include <ATen/CPUGenerator.h>
-
-#define const_generator_cast(generator) \
-  dynamic_cast<const CPUGenerator&>(generator)
+#include <c10/util/C++17.h>
+#include <algorithm>
+#include <random>
+#include <chrono>
 
 namespace at {
 
-CPUGenerator::CPUGenerator(Context * context_)
-  : context(context_), generator(THGenerator_new())
-{}
+namespace detail {
+
+// Ensures default_gen_cpu is initialized once.
+static std::once_flag cpu_device_flag;
+
+// Default, global CPU generator.
+static std::unique_ptr<CPUGenerator> default_gen_cpu;
+
+/**
+ * PyTorch maintains a collection of default generators that get
+ * initialized once. The purpose of these default generators is to
+ * maintain a global running state of the pseudo random number generation,
+ * when a user does not explicitly mention any generator.
+ * getDefaultCPUGenerator gets the default generator for a particular
+ * device.
+ */
+std::unique_ptr<CPUGenerator>& getDefaultCPUGenerator() {
```

Comment:
A lot of places, you take the returned `unique_ptr` and then immediately call `get()` on it. In fact, I audited most of the sites, and you don't seem to make use of the fact that you returned a mutable reference to a unique ptr at all. Why not return just a `CPUGenerator*` in that case? This is a better, narrower API, because it doesn't permit a user to rewrite the unique pointer to another generator living in different memory space.

Reply:
I think I kept it a mutable reference to a unique ptr so that I copy rng state of another generator to the default generator:
```
TEST(CPUGenerator, TestCloning) {
  // Test Description: 
  // Check copy assignment
  // See Note [Thread-safety and Generators]
  auto new_gen = at::detail::createCPUGenerator();
  new_gen->random(); // advance new gen_state
  new_gen->random();
  auto& default_gen = at::detail::getDefaultCPUGenerator();
  default_gen = new_gen->clone();
  ASSERT_EQ(new_gen->random(), default_gen->random());
}
```
My clone method wraps a unique ptr to the output of clone_impl():
```
template <typename Derived, typename Base>
struct CloneableGenerator : public Base {

  CloneableGenerator(Device device_in) : Base(device_in) {}
  virtual ~CloneableGenerator() = default;

  std::unique_ptr<Derived> clone() const {
    return std::unique_ptr<Derived>(static_cast<Derived*>(this->clone_impl()));
  }

private:
  virtual CloneableGenerator* clone_impl() const = 0;
};
```

Reply:
all `.get()` are with check_generator_with_default and in the torch/csrc/Generator.cpp. How about if I just change check_generator_with_default's signature to this:
```
template <typename T>
static inline T * check_generator_with_default(Generator * expr, std::unique_ptr<T>& defaultValue) {
  if (!expr) {
    expr = defaultValue.get();
  }
.
.
.
```

Reply:
> I think I kept it a mutable reference to a unique ptr so that I copy rng state of another generator to the default generator:

I don't see why this implies you need a unique pointer. If you want to copy assign a `T* x`, you can just say `*x = ...` I don't see anything in the example you posted that implies that a unique ptr must be used; maybe you can point it out more explicitly to me?

Reply:
I can't do `*x = ...` for generator class since copy assignment is deleted? I am a little confused now. May be this helps? I needed clone() method to be unique_ptr so that the caller doesn't worry about delete.

```
#include <memory>
#include <iostream>

struct CPUGenerator {
    CPUGenerator(int a_in) : a_(a_in) {}
    CPUGenerator(const CPUGenerator&) = delete;
    CPUGenerator& operator=(const CPUGenerator&) = delete;
    CPUGenerator* clone_with_raw();
    std::unique_ptr<CPUGenerator> clone_with_unique();
    int a_;
};

CPUGenerator* CPUGenerator::clone_with_raw() {
  return new CPUGenerator(a_);
}

std::unique_ptr<CPUGenerator> CPUGenerator::clone_with_unique() {
  return std::make_unique<CPUGenerator>(a_);
}


int main() {
    static std::unique_ptr<CPUGenerator> default_cpu_gen = std::make_unique<CPUGenerator>(5);
    CPUGenerator* new_gen = new CPUGenerator(6);
    CPUGenerator* default_cpu_gen_raw = default_cpu_gen.get();
    //*default_cpu_gen_raw = *new_gen; //won't compile
    default_cpu_gen_raw = new_gen->clone_with_raw(); // caller has to explicity delete pointer after use
    default_cpu_gen = new_gen->clone_with_unique(); // caller doesn't worry about ptr
}
```

Reply:
Ignore the comments about copy constructor; I think I understand the issue now.

> so that I can do `default_cpu_gen = new_gen->clone();`

Aha! So my problem is, I don't see why you would ever want to do this. I agree you do this in this test:

```
TEST(CPUGenerator, TestCloning) {
  // Test Description: 
  // Check copy assignment
  // See Note [Thread-safety and Generators]
  auto new_gen = at::detail::createCPUGenerator();
  new_gen->random(); // advance new gen_state
  new_gen->random();
  auto& default_gen = at::detail::getDefaultCPUGenerator();
  default_gen = new_gen->clone();
  ASSERT_EQ(new_gen->random(), default_gen->random());
}
```

But I think the test is wrong. I don't think this is an operation you should support on the generator.

There's an important reason why you shouldn't support this operation: we want to give a guarantee that the (raw) pointer returned by  `getDefaultCPUGenerator` has static lifetime; that is to say, it is always live. If you allow overwriting the default generator as you have done in this test, this invariant is broken, and people with `Generator*` to the old default generator will be left with a pointer to uninitialized memory. That's bad!

Reply:
Ok cool. I'll make the changes to keep the static lifetime invariant for the default generator.